zwave_js_server/event.py

```diff
@@ -1,21 +1,42 @@
 """Event base classes for Z-Wave JS."""
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List
+
+
+@dataclass
+class Event:
+    """Represent an event."""
+
+    type: str
+    data: dict = field(default_factory=dict)
 
 
 class EventBase:
-    def __init__(self):
+    """Represent a base for event handling model classes."""
+
+    def __init__(self, protocol: Any) -> None:
         """Initialize event base."""
-        self._listeners = {}
+        self.protocol = protocol
+        self._listeners: Dict[str, List[Callable]] = {}
 
-    def on(self, event_name, callback):
+    def on(self, event_name: str, callback: Callable) -> Callable:
         listeners: list = self._listeners.setdefault(event_name, [])
         listeners.append(callback)
 
-        def unsub():
+        def unsub() -> None:
             if callback in listeners:
                 listeners.remove(callback)
 
         return unsub
 
-    def emit(self, event_name, data):
+    def emit(self, event_name: str, data: dict) -> None:
         for listener in self._listeners.get(event_name, []):
             listener(data)
+
+    def _handle_event_protocol(self, event: Event) -> None:
+        """Process an event based on event protocol."""
+        protocol_type = self.protocol(event.type)
+        protocol_handler: Callable = getattr(
+            self, f"handle_{protocol_type.name.lower()}"
+        )
+        protocol_handler(event)
```

Comment:
I think that we can do it without a protocol. We can just check for the existence of the methods.

```suggestion
    handler = getattr(self, f"handle_{event.type.replace(' ', '_')}", None)
    if handler is None:
        LOGGER.debug("Received unknown event")
        return
    handler(event)
```